provider/randomComponent.go

```diff
@@ -15,35 +15,34 @@
 package provider
 
 import (
+	"fmt"
+
 	"github.com/pulumi/pulumi-go-provider/infer"
 	"github.com/pulumi/pulumi-random/sdk/v4/go/random"
 	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
 )
 
 // Similar to resources, components have a controlling struct.
-// Components only have a `Construct` method, which is mandatory.
-// The Construct method is responsible for creating the component by composing together other resources.
-type RandomComponent struct{}
+// The NewRandomComponent function is responsible for creating
+// the component by composing together other resources.
+type RandomComponent struct {
+	pulumi.ResourceState                     // Component state needs this for tracking nested resource states.
+	RandomComponentArgs                      // Include all the input fields in the state.
+	Password             pulumi.StringOutput `pulumi:"password"`
+}
 
 // Similar to resources, components have an input struct, defining what arguments it accepts.
 type RandomComponentArgs struct {
 	Length pulumi.IntInput `pulumi:"length"`
 }
 
-// Components also have a state, describing the fields that exist on the created component.
-type RandomComponentState struct {
-	pulumi.ResourceState                     // Component state needs this for tracking nested resource states.
-	RandomComponentArgs                      // Include all the input fields in the state.
-	Password             pulumi.StringOutput `pulumi:"password"`
-}
-
-func (r *RandomComponent) Construct(ctx *pulumi.Context, name, typ string, args RandomComponentArgs, opts pulumi.ResourceOption) (*RandomComponentState, error) {
+func NewRandomComponent(ctx *pulumi.Context, name string, args RandomComponentArgs, opts ...pulumi.ResourceOption) (*RandomComponent, error) {
 	// Initialize the component state.
-	comp := &RandomComponentState{
+	comp := &RandomComponent{
 		RandomComponentArgs: args,
 	}
 	// Register the component resource to which we will attach all other resources.
-	err := ctx.RegisterComponentResource(typ, name, comp, opts)
+	err := ctx.RegisterComponentResource(fmt.Sprintf("%s:RandomComponent", Name), name, comp, opts...)
```

Comment:
Doesn't this need to have an `:index:` segment to be a valid type?
```suggestion
	err := ctx.RegisterComponentResource(Name + ":index:RandomComponent", name, comp, opts...)
```